### ISSUE TITLE
Help wanted: JSON sanity check tools

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#40]
### Discussed in https://github.com/ioccc-src/mkiocccentry/discussions/39

<div type='discussions-op-text'>

<sup>Originally posted by **lcn2** January 29, 2022</sup>
# We need help with more tools in this repo

You can help.  See below.

# We are working hard to be able to run a MOCK IOCCC in 2022

By MOCK IOCCC we mean a beta test run of an IOCCC where entries submitted would NOT be judged.  Instead the IOCCC judges would go though the motions of running mock contest in order to beta test the process.  Nobody would win the MOCK IOCCC, although we would certainly thank those who helped test the process with special thanks to those who found something in the MOCK process that we need to fix.

> Any important lessons learned during the MOCK IOCCC would result in changes to the real IOCCC28.

Our goal is to hold the MOCK IOCCC in 2022 and then start the reason IOCCC28 in 2022.

There is A LOT of work that goes on behind the scenes by the IOCCC judges.  You can help the IOCCC judges by helping us with the following tools related to mkiocccentry.  See below.

# JSON sanity check tools

- Write a tool that attempts validate (performs sanity checks) on the .info.json produced by mkiocccentry.
- Write a tool that attempts validate (performs sanity checks) on the .author.json produced by mkiocccentry

Those two tools could be written as pull requests to add them as separate programs to this repo.  Moreover mkiocccentry could be modified to perform a function call into that code (in the same way the code iocccsize function is executed directly from mkiocccentry while iocccsize is also compiled as a separate tool).

Even though mkiocccenty would run the core of those tools as function calls, we still need the tools as standalone programs for use elsewhere.  So both mkiocccentry callable functions AND standalone tools are needed.

The standalone tools should take a file as an argument. If sane, these tools should exit 0, otherwise exit non-zero.

# compressed tarball filename sanity check tool

- Write a tool that validates the filename of a compressed tarball file.

The mkiocccenty tool forms a compressed tarball file with a very specific form of a filename (for internal IOCCC judging process reasons).  We need a tool that performs a sanity check on the filename of a compressed tarball file.

That tool could be added to this repo as a pull request.  Moreover mkiocccentry could be modified to perform a function call into that code (in the same way the code iocccsize function is executed directly from mkiocccentry while iocccsize is also compiled as a separate tool) as a sanity check of what mkiocccentry produced.

Even though mkiocccenty would run the core of those tools as function calls, we still need the tool as standalone programs for use elsewhere.  So both mkiocccentry callable functions AND standalone tool are needed.

The standalone tool should take a file as an argument. If sane, these tool should exit 0, otherwise exit non-zero.

# compressed tarball sanity check tool

- Write a tool to validate a compressed tarball produced by mkiocccentry

Such a compressed tarball tool would need to perform safety and sanity checks on the compressed tarball WITHOUT uncompressing it on disk. It would need to be paranoid, expecting the worst while attempting to validate for the best.

For example, it could run "tar -t" and look at the output to verify that the files un-tar into a sub-directory (i.e., not wander off with ../../../foo path, not attempting to write something under /foo).  

It would also call the code of the "validate the filename of the compressed tarball file" (in the same way the code iocccsize function is executed directly from mkiocccentry while iocccsize is also compiled as a separate tool).

This tool would need to verify that the sub-directory that the compressed tarball would produce is related to the previously validated compressed tarball filename.

This compressed tarball tool would need to perform the two checks on the tarball size both checking the size of the compressed tarball and tar reports as the overall size of the elements of the tarball (see limit_ioccc.h). 

This sanity check tool should verify that the compressed tarball has the required files (prog.c, Makefile, remarks.md, .info.json, .author.json).

The sanity check tool should be sure that no other files beginning with "." would be created.  It should check that no sub-sub-directories would be created, just the entry directory from which the tarball was created in the first place.  It should check that all files created are UNDER the entry directory only.

That tool could be added to this repo as a pull request.  Moreover mkiocccentry could be modified to perform a function call into that code (in the same way the code iocccsize function is executed directly from mkiocccentry while iocccsize is also compiled as a separate tool) as a sanity check of what mkiocccentry produced.

Even though mkiocccenty would run the core of those tools as function calls, we still need the tool as standalone programs for use elsewhere.  So both mkiocccentry callable functions AND standalone tool are needed.

The standalone compress tarball sanity check tool should take a file as an argument. If sane, these tool should exit 0, otherwise exit non-zero.

# In general

The reason why mkiocccentry should be able to execute the sanity check code of the above mention tools is that we (the IOCCC judges) plan to use those standalone tools in our judging procedures.  Nevertheless mkiocccentry should perform the same sanity checks to be help be sure what is generates will later will be checked by the IOCCC judges via the standalone tools.  

We don't want a mismatch of what mkiocccentry produces to conflict with the result of the sanity check standalone tools.  We don't want someone's IOCCC entry they worked hard on, invalidated because mkiocccenty produced something that the standalone tool later rejected.  That is why mkiocccentry should call on the sanity check code to validate what it is doing / creating.

Instead of loading up mkiocccentry with fork/exec of calling standalone tools and parsing the results, also allow function calls from mkiocccentry to run the same sanity checking code.

These sanity check tools need to be written in a clean, highly readable, trivial to understand style.  Ironic with such a request given that this is in support of Obfuscated C, yes, but IOCCC is about irony in a way.

We want it to be easy for us and others to inspect what this code is doing.  We want it easy for this code to be modified and improved on in the future.

We want this code to stand as a good example of how to write well written, easy to understand, easy to improve C code.

Use of the dbg facility, and the -v level debugging style used by mkiocccentry will be important.  And if a similar coding style with ALL libc function calls checked for errors is also important.

And as a C contest, we would highly prefer these tools be written in portable C and compiled with gcc or clang without warnings.

# In summary

Your help wanted AND is appreciated!
</div>

**Comment by lcn2** [comment]
Please also see these issues where help is wanted:
See also:

- Help wanted: compressed tarball filename sanity check tool #41 
- Help wanted: compressed tarball sanity check tool #42 

Thanks for your consideration and help!

**Comment by ilyakurdyukov** [comment]
> These sanity check tools need to be written in a clean, highly readable, trivial to understand style.

I have my own JSON reader, but it doesn't meet this requirement for sure.

**Comment by lcn2** [comment]
> > These sanity check tools need to be written in a clean, highly readable, trivial to understand style.
> 
> I have my own JSON reader, but it doesn't meet this requirement for sure.

The sanity check tools do NOT need to be a general JSON parser.  They simply need to verify that the JSON file produced by mkiocccentry contains all of the required elements, that the values match the checks imposed by mkentry, the lengths are within the limits imposed on by limit_ioccc.h, etc.

**Comment by xexyl** [comment]
It looks like Ilya has played with JSON so maybe he'd like to do this. I've never played with it so it would be something new for me and better to have someone experienced in it. I however do have a question about this mock IOCCC.

Because there wouldn't be winners I imagine that we should not submit entries that we have written (I know Ilya has finished some entries for example) that we would want to win in the future. Is this correct? If yes what should we submit?

Also I guess it was a typo when you said you would have the real IOCCC28 in 2022 and that you actually meant 2023? This would be very convenient for me as I'm in a difficult place right now (though I'm happy I could find the motivation to contribute here).

Thank you for giving us this information! Am I correct in my questions?

**Comment by ilyakurdyukov** [comment]
Cody, we can send in the old entries from previous contests or just random code, it's just a test anyway.

**Comment by xexyl** [comment]
> Cody, we can send in the old entries from previous contests or just random code, it's just a test anyway.

That's a good thought! Thank you for the suggestion. I really had no idea what I would send in but your idea of previous entries sounds good. I know at why at least two of my previous entries did not win (too similar to previous winning entries) so I could just submit those - because as you say it's just a test.

I do hope it was a typo though and that the next real contest is in 2023 - that would give me a lot of time to get in a better state.

**Comment by ilyakurdyukov** [comment]
I want the contest to be as soon as possible.

**Comment by xexyl** [comment]
I know mate. I understand. In some ways I want it too but I could use more time too. 

I have been very overwhelmed lately. Having the contest open might actually make me feel some better (even if temporarily) just like contributing to the tool.

**Comment by lcn2** [comment]
> .... I however do have a question about this mock IOCCC.
> 
> Because there wouldn't be winners I imagine that we should not submit entries that we have written (I know Ilya has finished some entries for example) that we would want to win in the future. Is this correct? If yes what should we submit?

The MOCK IOCCC would be about testing the IOCCC contest process.  We don't want to have a big failure that would force people to re-submit using an updated tool.  We want to avoid losing people's hard work because we messed something up in the process.

With the MOCK IOCCC, if we found a serious problem, we could abort the MOCK IOCCC, fix stuff, and then ask people to re-submit.  We would prefer that the MOCK IOCCC run without having to do that.  Still if it has to happen, better to happen during a MOCK IOCCC than during the real IOCCC28.

So yes, we would not judge the MOCK IOCCC submissions, just go through the motions and thank those who helped us test the MOCK IOCCC.

> Also I guess it was a typo when you said you would have the real IOCCC28 in 2022 and that you actually meant 2023? 

It would be nice if we can get to the point where the MOCK IOCCC runs in 2022 and any lessons are applied before we run IOCCC28.

If there is time after all that in 2022, we might start IOCCC28.  Our guess is that if we meet that deadline then IOCCC28 might start very late in 2022 and stay open thru to the beginning of 2023.  That would probably be the earliest IOCCC28 would run.

If the start of IOCCC28 slid into 2023 it would have to start in mid-May 2023 as one of the judges (Landon) is an astronomer who will be managing an experiment during the 2023 total solar eclipse and will be offline for a while.

There are other reasons we might delay IOCCC28 until mid-May 2023.  Nevertheless we work to have a tool chain and process to hold a MOCK IOCCC in 2022 and see if we can meet the end of 2022 deadline.

We (the IOCCC judges) have NOT set a time for IOCCC28.  And there is enough up in the air now for us to know we shouldn't promise a start time for IOCCC28.

There are many who want is to start IOCCC28 now.  However we don't want to repeat the mistakes of IOCCC27 and IOCCC26 .. long delays and other problems.

>This would be very convenient for me as I'm in a difficult place right now (though I'm happy I could find the motivation to contribute here).

Sorry to here about your difficult place @xexyl.  Perhaps 2023 will being better places for all of us.

> Thank you for giving us this information! Am I correct in my questions?

Yes.

**Comment by lcn2** [comment]
> Cody, we can send in the old entries from previous contests or just random code, it's just a test anyway.

The IOCCC judges wouldn't look much at the code during the MOCK IOCCC.  We would just give things a glance to see that there is C code, a Makefile that does something, a remarks file that looks like something .. but not really read it.

As we do in a real IOCCC, we would NOT know who is submitting to the MOCK IOCCC.

At the end, we would "pretend that everyone won", process the .author.json files and thank those who helped in the MOCK IOCCC.

**Comment by lcn2** [comment]
> I know mate. I understand. In some ways I want it too but I could use more time too. I have been very overwhelmed lately. Having the contest open might actually make me feel some better (even if temporarily) just like contributing to the tool.
> […](#)

While our problems may not be a significant as your problems: we too are impacted.

Spending time and effort on IOCCC tools has been very helpful for us feeling less overwhelmed.  **And we DO APPRECIATE the efforts folks such as yourself have made to help!**

**Comment by xexyl** [comment]
> > .... I however do have a question about this mock IOCCC.
> > Because there wouldn't be winners I imagine that we should not submit entries that we have written (I know Ilya has finished some entries for example) that we would want to win in the future. Is this correct? If yes what should we submit?
> 
> The MOCK IOCCC would be about testing the IOCCC contest process. We don't want to have a big failure that would force people to re-submit using an updated tool. We want to avoid losing people's hard work because we messed something up in the process.

That sounds like a good idea then.

> 
> With the MOCK IOCCC, if we found a serious problem, we could abort the MOCK IOCCC, fix stuff, and then ask people to re-submit. We would prefer that the MOCK IOCCC run without having to do that. Still if it has to happen, better to happen during a MOCK IOCCC than during the real IOCCC28.

Good point. It sounds like a trial contest is the way to go even if that delays the actual contest.

> > Also I guess it was a typo when you said you would have the real IOCCC28 in 2022 and that you actually meant 2023?
> 
> It would be nice if we can get to the point where the MOCK IOCCC runs in 2022 and any lessons are applied before we run IOCCC28.
> 
> If there is time after all that in 2022, we might start IOCCC28. Our guess is that if we meet that deadline then IOCCC28 might start very late in 2022 and stay open thru to the beginning of 2023. That would probably be the earliest IOCCC28 would run.

Thanks. That helps.

> 
> If the start of IOCCC28 slid into 2023 it would have to start in mid-May 2023 as one of the judges (Landon) is an astronomer who will be managing an experiment during the 2023 total solar eclipse and will be offline for a while.
> 

I really hope you get to document your work and I hope you have a blast!

> There are other reasons we might delay IOCCC28 until mid-May 2023. Nevertheless we work to have a tool chain and process to hold a MOCK IOCCC in 2022 and see if we can meet the end of 2022 deadline.

That sounds good to me.

> 
> We (the IOCCC judges) have NOT set a time for IOCCC28. And there is enough up in the air now for us to know we shouldn't promise a start time for IOCCC28.
> 
> There are many who want is to start IOCCC28 now. However we don't want to repeat the mistakes of IOCCC27 and IOCCC26 .. long delays and other problems.

Personally that worked out well for me last time as you reopening the contest due to COVID gave me the opportunity to write the Enigma machine. Still it was unfortunate that the publication of the entries took so long.

What I really hope is that you get to judge in person next time and you have your cook who can make you the chocolate cake I included the recipe to! I was so hoping to see you tweet the pictures of it during the judging process. I might have an amendment to the recipe but I haven't confirmed that it is the case yet. I will know after my birthday (14 February). 

> 
> > This would be very convenient for me as I'm in a difficult place right now (though I'm happy I could find the motivation to contribute here).
> 
> Sorry to here about your difficult place @xexyl. Perhaps 2023 will being better places for all of us.
> 

Thank you very much. I really appreciate it! In some ways it might be but I'm doubtful of it with the anti-vaxxers. I guess we'll have to wait and see. I turn 40 on 14 February and I'm really upset I can't go out to celebrate it due to the pandemic. It's not a good time to be reaching a new decade. Even if I was okay eating out during the pandemic (and I'm not) there's no way I would on my birthday - 14 February will be a super spreader day throughout the western world and that truly terrifies me!

But that's not the real problem I'm having - it just adds to it.


> > Thank you for giving us this information! Am I correct in my questions?
> 
> Yes.

Thank you very much for confirming.

**Comment by lcn2** [comment]
> What I really hope is that you get to judge in person next time and you have your cook who can make you the chocolate cake I included the recipe to! I was so hoping to see you tweet the pictures of it during the judging process. I might have an amendment to the recipe but I haven't confirmed that it is the case yet. I will know after my birthday (14 February).

We could make cakes in parallel, at least for those judges that are cooks.  :-)

**Comment by xexyl** [comment]
> > What I really hope is that you get to judge in person next time and you have your cook who can make you the chocolate cake I included the recipe to! I was so hoping to see you tweet the pictures of it during the judging process. I might have an amendment to the recipe but I haven't confirmed that it is the case yet. I will know after my birthday (14 February).
> 
> We could make cakes in parallel, at least for those judges that are cooks. :-)

That's true. I wish you had done that! It would have felt like another victory :) But it can be a daunting recipe to be sure. It takes hours to make and bake. Still if you love chocolate it's very well worth it! My late stepmum only liked white chocolate and she loved this cake! On the other hand my late grandfather never would have had this cake as he was crazy and didn't like chocolate (though depending on how chocolaty they are he was okay with brownies, somehow).

Still I hope you get to try it out the next time..whenever you'll have judging in person. This does remind me that I need to do pull requests for my winning entries in 2020 but these seem more important. If the amending of the recipe has to be done I'll bring it up for sure - at least on twitter but eventually the markdown files.

**Comment by xexyl** [comment]
> > Cody, we can send in the old entries from previous contests or just random code, it's just a test anyway.
> 
> The IOCCC judges wouldn't look much at the code during the MOCK IOCCC. We would just give things a glance to see that there is C code, a Makefile that does something, a remarks file that looks like something .. but not really read it.
> 
> As we do in a real IOCCC, we would NOT know who is submitting to the MOCK IOCCC.
> 
> At the end, we would "pretend that everyone won", process the .author.json files and thank those who helped in the MOCK IOCCC.

I was more thinking that if I included an entry I already wrote, say, that when the real contest comes along then it would be harder for this entry to win. But as Ilya pointed out I can just include old code or something else entirely.

**Comment by xexyl** [comment]
> > I know mate. I understand. In some ways I want it too but I could use more time too. I have been very overwhelmed lately. Having the contest open might actually make me feel some better (even if temporarily) just like contributing to the tool.
> > […](#)
> 
> While our problems may not be a significant as your problems: we too are impacted.
> 
I wouldn't go there mate! We all have our own issues at times. It doesn't matter if my situation is worse than yours or not; what matters is we all are impacted during this time and even if what I referred to is not to do with the pandemic part of my situation IS to do with the pandemic. I appreciate the thought but I don't want to make it seem like my problems are more important or worse. They're just different.

> Spending time and effort on IOCCC tools has been very helpful for us feeling less overwhelmed. **And we DO APPRECIATE the efforts folks such as yourself have made to help!**

Thank you very much for the kind words! I'm happy I can help. And now I'm really going for the night. Take good care and stay safe! I'll read more tomorrow I hope.

**Comment by lcn2** [comment]
Is there a JSON sanity check and JSON parsing library, callable in C, that someone can recommend?

This JSON sanity check tool might not be part of the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), so it doesn't have to be part of a Unix distribution.  But if it is part of a Homebrew package, or a well maintained GitHub repo that would be OK.

**Comment by xexyl** [comment]
I was going to say that after I get `txzchk` sorted out (which seems like it might be soon) I was thinking about looking into learning about JSON and maybe coming up with something (I'd have to look at the requirements again though). 

However I want to caution about a third party library IF the tool has to be invoked by submitters. The reason is different prefixes. For example under `MacPorts` the prefix is `/opt/local`. I personally don't use Homebrew because at least at one point they take ownership of `/usr/local` which is a serious security risk. For example Sophos actually writes to that directory but with Homebrew it no longer is owned by root. Supposedly they have fixed this but I think they still initially have to do something with it when the obvious correct choice would be to use sudo (which is how you install things with MacPorts).


> Is there a JSON sanity check and JSON parsing library, callable in C, that someone can recommend?
> 
> This JSON sanity check tool might not be part of the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), so it doesn't have to be part of a common Unix distribution. But if it is part of a Homebrew package, or a well maintained GitHub repo that would be OK.

On the other hand if it was a library that was compiled in the IOCCC tool set it might be okay since you could just specify the library include path (I mean `-L` and also `-I` options to the compiler).

Anyway I hope I can finish the other tool tomorrow or if not then Friday. Saturday I'll be free but most of Sunday I won't be; Monday I'll be pretty busy too and part of Tuesday as well (I will say more about Sunday/Monday when the time comes if I am in the middle of working on anything).

If I can I'll look into JSON in the near future.

**Comment by lcn2** [comment]
> I was going to say that after I get `txzchk` sorted out (which seems like it might be soon) I was thinking about looking into learning about JSON and maybe coming up with something (I'd have to look at the requirements again though).
> 
> However I want to caution about a third party library IF the tool has to be invoked by submitters. The reason is different prefixes. For example under `MacPorts` the prefix is `/opt/local`. I personally don't use Homebrew because at least at one point they take ownership of `/usr/local` which is a serious security risk. For example Sophos actually writes to that directory but with Homebrew it no longer is owned by root. Supposedly they have fixed this but I think they still initially have to do something with it when the obvious correct choice would be to use sudo (which is how you install things with MacPorts).
> 
> > Is there a JSON sanity check and JSON parsing library, callable in C, that someone can recommend?
> > This JSON sanity check tool might not be part of the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry), so it doesn't have to be part of a common Unix distribution. But if it is part of a Homebrew package, or a well maintained GitHub repo that would be OK.
> 
> On the other hand if it was a library that was compiled in the IOCCC tool set it might be okay since you could just specify the library include path (I mean `-L` and also `-I` options to the compiler).
> 
> Anyway I hope I can finish the other tool tomorrow or if not then Friday. Saturday I'll be free but most of Sunday I won't be; Monday I'll be pretty busy too and part of Tuesday as well (I will say more about Sunday/Monday when the time comes if I am in the middle of working on anything).
> 
> If I can I'll look into JSON in the near future.

We plan to move this issue to a new repository.  In that repo we will produce a more detailed set of specs.

We prefer `Homebrew` over `Macports` for various reason, FYI.

Stay tuned.

**Comment by xexyl** [comment]
> We plan to write up two separate requests for checking `.info.json` files and for checking `.author.json` files, and then close down this request in favor of more explicit requests.
> 

Once `txzchk` is in order I will start reading up on JSON. Maybe I can do a bit of reading in between too but we shall see. 

> We would prefer to NOT have to depend on someone external JSON file checker for this [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry). As such, the two JSON checks cannot link to, or depend some external JSON parser.
> 

I strongly agree with this. It would be better this way because if it’s going to have to be used by contestants then it should be something that everyone can use without requiring them to install additional libraries etc. 

> Perhaps what can be done is to assume that the format of `.info.json` files and of `.author.json` files is for format that `mkiocccentry` creates, instead of trying to do some general JSON file parser code.

That’s a good idea and probably the better way to go about it. 

> 
> For example, the `.info.json` file could be read into memory and then `strtok(3)` could be used to break up the file into JSON tokens. Then those tokens could be examined for the required set of tokens, the version format string, presence of unknown tokens, etc.?

I was actually thinking about something like this. Do you plan on writing more specific details? That would be helpful I think. 

I have enjoyed working on the other tool so much that this sounds like something I would enjoy too!

**Comment by xexyl** [comment]
I just started working on this but without knowing the full details it's only the parsing of options so far. I might (later on today when I find more energy - else tomorrow) start the parsing i.e. extraction of the fields in the files (can parse both `.info.json` and `.author.json` - but only one at a time) but for now this is what's done:

- Since the fields that belong in the two json files are in structs in `mkiocccentry.c` I have moved those two structs into a separate header file `json.h` (the name seemed appropriate).
- Added `#include json.h` to both `mkiocccentry.c` and the new file (which for the time being I called `jsonchk.c`).
- Updated the Makefile to compile this.

Additionally `main()`:

- parses the args (see below)
- verifies that the file specified exists and is a regular file and is readable via sanity check function
- exits depending on any issues found

As for the options right now:

```shell
$ ./jsonchk -h
-h help mode

usage: ./jsonchk [-h] [-v level] [-V] [-i] [-a] jsonpath

	-h			print help message and exit 0
	-v level		set verbosity level: (def level: 0)
	-V			print version string and exit

	-i			parse .info.json file
	-a			parse .author.json file

	jsonpath		path to an IOCCC JSON file

jsonchk version: 0.1 2022-02-11
```

More obviously has to be done but this is a good start I think. If you like I can do a pull request - or make modifications (e.g. to the name or whatever else) and then do that. 

If I do do this though I should change the `struct info` in `txzchk` to be named differently perhaps `txzchk_info` or `tarball_info` even though the file wouldn't include `json.h` - because it might be ambiguous to some (and it would make it more explicit). 

Just let me know.

I do realise that you're planning on opening a new issue but since that issue is not yet open I figured I'd add my comments here.

**Comment by lcn2** [comment]
This issue has been replaced by issues #68 and #69.
